js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+var flexbuffers = {};
+
+flexbuffers.BitWidth = {
+  WIDTH8: 0,
+  WIDTH16: 1,
+  WIDTH32: 2,
+  WIDTH64: 3,
+};
+
+flexbuffers.BitWidth.toByteWidth = (bitWidth) => {
+  return 1 << bitWidth;
+};
+
+flexbuffers.BitWidth.width = (value) => {
+  if (Number.isInteger(value)) {
+    const v = value < 0 ? value * -1 : value;
+    if (v >> 7 === 0) return flexbuffers.BitWidth.WIDTH8;
+    if (v >> 15 === 0) return flexbuffers.BitWidth.WIDTH16;
+    if (v >> 31 === 0 && v >> 32 === v) return flexbuffers.BitWidth.WIDTH32;
+    return flexbuffers.BitWidth.WIDTH64;
+  }
+
+  function _toF32(value) {
+    const buffer = new ArrayBuffer(4);
+    const view = new DataView(buffer);
+    view.setFloat32(0, value, true);
+    return view.getFloat32(0, true);
+  }
+
+  return value === _toF32(value) ? flexbuffers.BitWidth.WIDTH32: flexbuffers.BitWidth.WIDTH64;
+};
+
+flexbuffers.BitWidth.fwidth = (value) => {
+  function _toF32(value) {
+    const buffer = new ArrayBuffer(4);
+    const view = new DataView(buffer);
+    view.setFloat32(0, value, true);
+    return view.getFloat32(0, true);
+  }
+
+  return value === _toF32(value) ? flexbuffers.BitWidth.WIDTH32: flexbuffers.BitWidth.WIDTH64;
+};
+
+flexbuffers.BitWidth.uwidth = (value) => {
+  if (Number.isInteger(value) && value >= 0) {
+    if (value >> 8 === 0) return flexbuffers.BitWidth.WIDTH8;
+    if (value >> 16 === 0) return flexbuffers.BitWidth.WIDTH16;
+    if (value >> 32 === 0) return flexbuffers.BitWidth.WIDTH32;
+    return flexbuffers.BitWidth.WIDTH64;
+  }
+
+  throw `Value: ${value} is not a uint`;
+};
+
+flexbuffers.BitWidth.fromByteWidth = (value) => {
+  if (value === 1) return flexbuffers.BitWidth.WIDTH8;
+  if (value === 2) return flexbuffers.BitWidth.WIDTH16;
+  if (value === 4) return flexbuffers.BitWidth.WIDTH32;
+  if (value === 8) return flexbuffers.BitWidth.WIDTH64;
+  throw "Unexpected value " + value;
+};
+
+flexbuffers.BitWidth.paddingSize = (bufSize, scalarSize) => {
+  return (~bufSize + 1) & (scalarSize - 1);
+};
+
+flexbuffers.ValueType = {
```

Comment:
Is this a map and does that involve a lookup every time we use these values? Are there more efficient ways to define constants in JS? remember that speed is important for this serialization format, so writing JS to suit that is important.

Reply:
It's fine, property lookups on static objects like this are translated to very efficient code and not hashmap-like lookups.

Reply:
Although I'd avoid defining methods (below) on the same object as these constants.

Reply:
> Although I'd avoid defining methods (below) on the same object as these constants.
@RReverser for performance reasons, or stylistic? Both is ok, just wondering.

Reply:
Both, really.

Reply:
I separated the values from functions now.
Now we have `flexbuffers.ValueType` and `flexbuffers.ValueTypeUtil`